src/PowerShellEditorServices/Language/FoldingOperations.cs

```diff
@@ -0,0 +1,65 @@
+//
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+//
+
+using System;
+using System.Collections.Generic;
+using System.Management.Automation.Language;
+
+namespace Microsoft.PowerShell.EditorServices
+{
+    /// <summary>
+    /// Provides common operations for code folding in a script
+    /// </summary>
+    internal static class FoldingOperations
+    {
+        /// <summary>
+        /// Extracts all of the unique foldable regions in a script given a script AST and the list tokens
+        /// used to generate the AST
+        /// </summary>
+        internal static FoldingReference[] FoldableRegions(
+            Token[] tokens,
+            Ast scriptAst,
+            bool ShowLastLine)
+        {
+            List<FoldingReference> foldableRegions = new List<FoldingReference>();
+
+            // Add regions from AST
+            foldableRegions.AddRange(Microsoft.PowerShell.EditorServices.AstOperations.FindFoldsInDocument(scriptAst));
+
+            // Add regions from Tokens
+            foldableRegions.AddRange(Microsoft.PowerShell.EditorServices.TokenOperations.FoldableRegions(tokens));
+
+            // Remove any null entries. Nulls appear if the folding reference is invalid or missing
+            foldableRegions.RemoveAll(item => item == null);
+
+            // Sort the FoldingReferences, starting at the top of the document,
+            // and ensure that, in the case of multiple ranges starting the same line,
+            // that the largest range (i.e. most number of lines spanned) is sorted
+            // first. This is needed to detect duplicate regions. The first in the list
+            // will be used and subsequent duplicates ignored.
+            foldableRegions.Sort();
+
+            // It's possible to have duplicate or overlapping ranges, that is, regions which have the same starting
+            // line number as the previous region. Therefore only emit ranges which have a different starting line
+            // than the previous range.
+            foldableRegions.RemoveAll( (FoldingReference item) => {
```

Comment:
Maybe spin this into a static method called something like `HasParentWithSameStartLine`. Easier to reason about while reading down the page and self-documenting

Reply:
This is only 4 lines of code.  Not sure it really warrants to be extracted into it's own method.

Reply:
Some principles at work in my mind are:
- We need some threshold for saying an inline lambda should be a new method. I think [Python's builtin threshold](https://stackoverflow.com/questions/1233448/no-multiline-lambda-in-python-why-not) is as good a choice as any: more than one line. That prevents the temptation to add complexity to existing inline lambdas. My personal feeling is that any lambda requiring braces should be a static method.
- Reasoning about control flow while reading down the listing is harder with a lambda. It's easy to make mistakes about what statements happen at what time for want of a brace. There's a reason they call it "callback hell".
- It's a clearer signal to the C# compiler that you don't need any local variables lifted into the closure. I don't think it's a significant effect currently, but still think of it as a pro.

Reply:
I'll defer to @TylerLeonhardt's opinion on this one though

Reply:
@glennsarti + @rjmholt: what do you think of this implementation:
https://gist.github.com/TylerLeonhardt/48212963362d4e0e38f89aa2e807db62

it's a little longer in terms of lines but as far as allocation and time complexity it's a tad better I think:

Allocates: 
1 Dictionary,
1 Array (from .Values of Dictionary)

Runtime complexity:
O(n + nlogn) (because of the Sort)

The current implementation allocs:
1 List
1 List for the RemoveAll
1 Array (via the ToArray)
1 delegate

with a runtime complexity of:
O( n + nlogn + nlogn) 

nlogn for Sort
nlogn for BinarySearch (logn) inside of RemoveAll (n)

at the end of the day, this is certainly not a HUGE boost, but micro-optimizations will benefit us since we're currently experiencing high perf related to garbage collection from allocations and general unoptimized code.

Reply:
@TylerLeonhardt Alternate solution to your gist - https://github.com/PowerShell/PowerShellEditorServices/pull/806/commits/e2dc02b4f08224ef326fdf3a9c543bbc45524545

I need to cleanup the commit message but;

Instead of using a intermediate arrays, extend a Dictionary object with a `SafeAdd` method which encapsulates the rules for adding a reference.  And then pass that through the other methods using `ref`.  Kinda like an accumulator pattern.  And then only serialize to an array once in the Language Server.

Your method (I think) actually allocates 
1 Dictionary,
2 Lists (One in each of the AST and Token operations)
3 Array (from .Values of Dictionary) (temporary arrays when collating the AST and Token operations)

Whereas mine is just 1 Dictionary

(We both exclude the allocations in the LanguageServer.cs code)

Reply:
I've decided to use my approach, and it's a separate commit in this PR.

Fixed.